fs/fs.go

```diff
@@ -74,20 +74,68 @@ import (
 	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
 )
 
 var (
 	defaultIndexSelectionPolicy = SelectFirstPolicy
 	fusermountBin               = "fusermount"
+	layerQueueBufferSize        = 1024
 )
 
+type layerQueue struct {
+	queue chan func() string
+	cache *sync.Map
+	smp   *semaphore.Weighted
+}
+
+func newLayerQueue(maxConcurrency int64) *layerQueue {
+	lq := &layerQueue{}
+	lq.queue = make(chan func() string, layerQueueBufferSize) // arbitrarily chosen buffer size
+	lq.cache = &sync.Map{}
+	lq.smp = semaphore.NewWeighted(maxConcurrency)
+	return lq
+}
+
+func (lq *layerQueue) Start(ctx context.Context) error {
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		for {
+			lq.smp.Acquire(egCtx, 1)
+
+			resolveFn := <-lq.queue
+			go func() {
+				digest := resolveFn()
+				lq.cache.Delete(digest)
+				lq.smp.Release(1)
+			}()
+		}
+	})
+
+	return nil
+}
+
+func (lq *layerQueue) Enqueue(digest string, fn func() string) {
```

Comment:
We currently reference a `layer` by it's [img ref + digest](https://github.com/awslabs/soci-snapshotter/blob/04a83aaa56243152a0c14828a12cb810997ea6d8/fs/layer/layer.go#L234), and so a `layer` is heavily associated with the image it's a part of. If we only map by digest, we'll only concurrently resolve one layer if two images share the same layer. We may want to make the key ref + digest, so that we can concurrently pre-resolve shared layers as-well.